notifier: fix notifications for some storage types, add tests

heikju opened the pull request:
For Azure, Swift and SFTP object storages, the attempt to get object size for completion notifications was incorrect, calling `os.path.getsize()` with a file-like object, instead of a path.

Given that the object to store may not be a file on disk, that can be seeked or `stat()`ed, we need some other way to obtain the number of bytes.

For these storages, the store call doesn't seem to return the object size in a convenient way, so make use of the progress callback we implement for object storage backends. That gives us the number of bytes transferred, that we can then pass on to notifier.

For tests, start with these three backends that seemed to have been broken by #67. Some libraries used by them are not in our requirements.dev.txt, so settle for mocking those imports.

<!-- All contributors please complete these sections, including maintainers -->
Looks to fix a few storage backends broken by #67, and adds new tests

<!-- Provide a small sentence that summarizes the change. -->

<!-- Provide the issue number below if it exists. -->
Resolves: #xxxxx

# Why this way

<!-- Provide a small explanation on why this is the approach you took for solving this problem. -->

kmichel-aiven commented on rohmu/object_storage/swift.py:
```diff
@@ -297,17 +297,24 @@ def copy_file(self, *, source_key, destination_key, metadata=None, **_kwargs):
             fd,
             cache_control=cache_control,
             metadata=metadata,
             mimetype=mimetype,
-            upload_progress_fn=upload_progress_fn,
+            upload_progress_fn=upload_progress_fn_wrapper,
             multipart=True,
             content_length=metadata.get("Content-Length"),
         )
-        self.notifier.object_created(key=key, size=os.path.getsize(fd))
+        self.notifier.object_created(key=key, size=notify_size[0])
```
Since the size will only be available when metadata.get("Content-Length") is available, why not using that directly instead of wrapping the callback just to get back the same value ?

kmichel-aiven commented on rohmu/object_storage/sftp.py:
```diff
@@ -192,15 +192,17 @@ def store_file_from_memory(self, key, memstring, metadata=None, cache_control=No
         self.log.debug("Store path: %r", target_path)
 
         def wrapper_upload_progress_fn(bytes_written, total_bytes):  # pylint: disable=unused-argument
+            notify_size[0] = bytes_written
```
This should accumulate the written bytes, not overwrite the previous value.

heikju replied:
Really? Paramiko docs say that the first arg is the number of bytes sent so far, and callback implementation here https://github.com/paramiko/paramiko/blob/main/paramiko/sftp_client.py#L675 seems to agree with that, already doing the accumulation of individual writes.

kmichel-aiven commented on rohmu/object_storage/azure.py:
```diff
@@ -345,11 +345,14 @@ def store_file_object(self, key, fd, *, cache_control=None, metadata=None, mimet
-            if upload_progress_fn and bytes_sent:
-                upload_progress_fn(bytes_sent)
+            if bytes_sent:
+                notify_size[0] = bytes_sent
```
This should accumulate the written bytes, not overwrite the previous value.

heikju replied:
It seems to me that the Azure client already accumulates this internally, though it turns out to be harder to find documentation about this, I wonder how recent the version we're using is. In any case, see line 301 of `/usr/lib/python3.10/site-packages/azure/storage/blob/_shared/policies.py`

kmichel-aiven commented on rohmu/object_storage/swift.py:
```diff
@@ -297,17 +297,24 @@ def copy_file(self, *, source_key, destination_key, metadata=None, **_kwargs):
+        notify_size = [0]
+
+        def upload_progress_fn_wrapper(bytes_written):
+            notify_size[0] = bytes_written
```
If this is kept (cf. comment below), it should accumulate the written bytes, not overwrite the previous value.

heikju replied:
It seems to me that this is called with the value content-length minus number of bytes not yet transferred, unless I misunderstand the logic in `_store_file_contents()`

kmichel-aiven replied:
Same as the previous two :+1:

alanfranz commented on test/test_object_storage_sftp.py:
```diff
@@ -0,0 +1,55 @@
+            transfer.store_file_from_disk(key="test_key1", filepath=tmpfile.name)
+
+        client.putfo.assert_called()
+        notifier.object_created.assert_any_call(key="test_key1", size=len(test_data))
```
I think this assertion is wrong. I'd assume (but you can prove me wrong) that we get 1 notification for each 1 object created, so this should assert that method 'object_created' is called once.

But I see that sftp store_file_from_disk calls store_file_object and both of them perform a notification, so there could be a bug there.

heikju replied:
Yes, we can use `assert_called_once_with()` everywhere to be more strict, no reason really to not do so.

alanfranz reviewed (dismissed):
I spotted a problem, I can't finish the whole review though right now.

kmichel-aiven commented on test/test_object_storage_azure.py:
```diff
@@ -0,0 +1,77 @@
+    get_blob_client_mock = MagicMock()
+    blob_client = MagicMock(get_blob_client=get_blob_client_mock)
+    service_client = MagicMock(from_connection_string=MagicMock(return_value=blob_client))
+    sys.modules["azure.common"] = MagicMock()
+    sys.modules["azure.core.exceptions"] = MagicMock()
+    sys.modules["azure.storage.blob"] = MagicMock(BlobServiceClient=service_client)
```
This manually alters global variables in a remote module and does not revert it at the end of the test.
This should use something like [mock.patch](https://docs.python.org/dev/library/unittest.mock.html#unittest.mock.patch).

heikju replied:
`patch.dict()` seems like the simplest way to do the trick.

kmichel-aiven commented on test/test_object_storage_azure.py:
```diff
@@ -0,0 +1,77 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
```
Please add parameter and return type hints on all new code.

kmichel-aiven commented on test/test_object_storage_swift.py:
```diff
@@ -0,0 +1,56 @@
+
+@pytest.fixture(scope="module", name="swift_module")
+def fixture_swift_module():
+    sys.modules["swiftclient"] = MagicMock()
```
Same as azure, this should use `mock.patch`

kmichel-aiven commented on test/test_object_storage_swift.py:
```diff
@@ -0,0 +1,56 @@
+    test_data = b"test-data"
+    file_object = BytesIO(test_data)
+
+    transfer.store_file_object(key="test_key2", fd=file_object, metadata={"Content-Length": len(test_data)})
```
The case without `Content-Length` should be either fixed to work like the other or be tested, separately.

kmichel-aiven reviewed (changes requested):
We found a bug in these three storages by reading their code, but all the storages modified in #67 should be tested.

heikju commented:
I hope I've addressed the comments to a sufficient degree. In the interest of making reviewing easier, I'd like to leave tests for the remaining storage backends to a follow-up PR.

heikju commented:
Updated to make type hints acceptable for Python 3.7

hackaugusto reviewed (approved):
LGTM, only had a small comment.
